x/gamm/twap/logic.go

```diff
@@ -0,0 +1,149 @@
+	// thus it is treated as the effective spot price until the new time.
+	// (As there was no change until at or after this time)
+	// TODO: Think about overflow
+	newRecord.P0ArithmeticTwapAccumulator = newRecord.P0ArithmeticTwapAccumulator.Add(
+		types.SpotPriceTimesDuration(record.P0LastSpotPrice, timeDelta))
+	newRecord.P1ArithmeticTwapAccumulator = newRecord.P1ArithmeticTwapAccumulator.Add(
+		types.SpotPriceTimesDuration(record.P1LastSpotPrice, timeDelta))
+	return newRecord
+}
+
+func (k Keeper) pruneOldTwaps(ctx sdk.Context) {
+	// TODO: Read this from parameter
+	lastAllowedTime := ctx.BlockTime().Add(-48 * time.Hour)
+	k.pruneRecordsBeforeTime(ctx, lastAllowedTime)
+}
+
+// getInterpolatedRecord returns a record for this pool, representing its accumulator state at time `t`.
+// This is achieved by getting the record `r` that is at, or immediately preceding in state time `t`.
+// To be clear: the record r s.t. `t - r.Time` is minimized AND `t >= r.Time`
+func (k Keeper) getInterpolatedRecord(ctx sdk.Context, poolId uint64, t time.Time, assetA, assetB string) (types.TwapRecord, error) {
+	if !(assetA > assetB) {
+		assetA, assetB = assetB, assetA
+	}
+	record, err := k.getRecordAtOrBeforeTime(ctx, poolId, t, assetA, assetB)
+	if err != nil {
+		return types.TwapRecord{}, err
+	}
+	record = interpolateRecord(record, t)
+	return record, nil
+}
+
+func (k Keeper) getMostRecentRecord(ctx sdk.Context, poolId uint64, assetA, assetB string) (types.TwapRecord, error) {
+	if !(assetA > assetB) {
+		assetA, assetB = assetB, assetA
+	}
+	record, err := k.getMostRecentRecordStoreRepresentation(ctx, poolId, assetA, assetB)
+	if err != nil {
+		return types.TwapRecord{}, err
+	}
+	record = interpolateRecord(record, ctx.BlockTime())
+	return record, nil
+}
+
+// interpolate record updates the record's accumulator values and time to the interpolate time.
+//
+// pre-condition: interpolateTime >= record.Time
+func interpolateRecord(record types.TwapRecord, interpolateTime time.Time) types.TwapRecord {
+	return recordWithUpdatedAccumulators(record, interpolateTime)
+}
+
+// TODO: Test math, test p0 vs p1 correctness
+// precondition: endRecord.Time > startRecord.Time
+func computeArithmeticTwap(startRecord types.TwapRecord, endRecord types.TwapRecord, quoteAsset string) sdk.Dec {
+	timeDelta := endRecord.Time.Sub(startRecord.Time)
+	// if time difference is 0, then return the last spot price based off of start.
+	if timeDelta == time.Duration(0) {
+		if quoteAsset == startRecord.Asset0Denom {
+			return endRecord.P0LastSpotPrice
+		}
+		return endRecord.P1LastSpotPrice
+	}
+	var accumDiff sdk.Dec
+	if quoteAsset == startRecord.Asset0Denom {
+		accumDiff = endRecord.P0ArithmeticTwapAccumulator.Sub(startRecord.P0ArithmeticTwapAccumulator)
+	} else {
+		accumDiff = endRecord.P1ArithmeticTwapAccumulator.Sub(startRecord.P1ArithmeticTwapAccumulator)
+	}
+	return types.AccumDiffDivDuration(accumDiff, timeDelta)
```

Comment:
I don't think we have a test for `AccumDiffDivDuration`. I think every exported function deserves a dedicated test. Please consider adding it

Reply:
True, alright if I add this in an upcoming PR as we add overflow and alternate duration handling?

For now correctness is tested by the TWAP numbers